container rename support

strzinek opened the pull request:
Allow to rename a container, if it is not running or paused.

There is an issue with container not being refreshed after renamed, as no rename event is fired. This should be fixed in podman v4.1 according to https://github.com/containers/podman/issues/13919

martinpitt commented:
Thanks @strzinek for all your work! I just wanted to explain the delays -- most of us had some holidays in the last week. We'll go through your PRs one by one, but they will reach require some work: All features, fixes, and other changes need to be covered with corresponding tests (functional, and where appropriate, the "pixel tests", i.e. screenshots). E.g. for this PR, testing the renaming should be added to `_testLifecycleOperations()`, or a new `testRename()`. There will likely be some fallout on some operating systems, which we'll need to handle.

Do you want to give it a try to write a test? Please start with trying to [run the existing ones](https://github.com/cockpit-project/cockpit-podman/blob/main/HACKING.md#running-tests-locally), to make sure you have everything set up. Then you can add a new one for renaming.

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+
+const _ = cockpit.gettext;
+
+class ContainerRenameModal extends React.Component {
```
In new code we try to use functional components and `useState()`, see [React hooks docs](https://reactjs.org/docs/hooks-state.html). If you don't feel comfortable with them, we are happy to help with this.

strzinek replied:
Fixed in commit https://github.com/cockpit-project/cockpit-podman/pull/966/commits/0c3706e1009fabc1fee2a17b247175993469a9c4

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+        super(props);
+        this.state = {
+            name: dockerNames.getRandomName(),
+            nameError: "",
```
"empty" values should be `null`.

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+    constructor(props) {
+        super(props);
+        this.state = {
+            name: dockerNames.getRandomName(),
```
The dialog input should start with the *current* container name. Often you merely want to tweak it a bit (perhaps add a suffix). I don't think we should use random names here at all -- the whole point of renaming is to give it some descriptive user-selected name.

jelly replied:
I agree, starting editing with the old container name makes more sense to me.

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    componentDidMount() {
+        this._isMounted = true;
```
`_isMounted` is not used anywhere. Please drop this function and the assignment in WillUnmount().

strzinek replied:
Fixed in commit https://github.com/cockpit-project/cockpit-podman/commit/0c3706e1009fabc1fee2a17b247175993469a9c4

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+        const newState = { [targetName]: value };
+
+        if (targetName === "name")
+            newState.nameError = "";
```
`null`

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+            return;
+        }
+
+        const renameData = {};
+        renameData.name = this.state.name;
```
Please merge into the initialization.

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+                .then(() => this.props.onHide())
+                .catch(ex => {
+                    this.setState({
+                        dialogError: cockpit.format(_("Failed to rename container $0"), this.props.container.Names),
```
"Names" sounds like this is a list, not a string. Pick the first element to be safe.

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+                               validated={this.state.nameError ? "error" : "default"}
+                               onChange={value => this.handleInputChange("name", value)} />
+                </FormGroup>
+
```
stray empty line

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+            <Modal isOpen
+                   showClose={false}
+                   position="top" variant="medium"
+                   title={cockpit.format(_("Rename container $0 "), this.props.container.Names)}
```
Please drop the trailing space in the string

martinpitt commented on src/Containers.jsx:
```diff
@@ -112,6 +114,15 @@ const ContainerActions = ({ container, onAddNotification, version, localImages }
     };
 
+    const renameContainer = () => {
+        if (container.State == "running") {
+            setRenameModal(false);
+        } else {
+            setRenameModal(true);
+        }
```
`setRenameModal(container.State !== "running")`

martinpitt commented on src/client.js:
```diff
@@ -98,6 +98,14 @@ export function delContainer(system, id, force) {
     });
 }
 
+export function renameContainer(system, id, config) {
+    return new Promise((resolve, reject) => {
+        podmanCall("libpod/containers/" + id + "/rename", "POST", config, system)
+                .then(resolve)
+                .catch(reject);
+    });
+}
```
I know this is just copying the previous pattern, but that Promise() wrapper is completely useless. Please use something like this:

```js
export const renameContainer = (system, id, config) => podmanCall("libpod/containers/" + id + "/rename", "POST", config, system);
```

or if you prefer, just change the function body to
```js
return podmanCall("libpod/containers/" + id + "/rename", "POST", config, system);
```

strzinek replied:
Thanks @martinpitt for the review, I will fix the code and prepare the tests when I have some free time to code. In my case it's usually during the holidays. ;-)

martinpitt reviewed (changes requested):
Thanks!

jelly commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,108 @@
+
+    handleRename() {
+        if (!this.state.name) {
+            this.setState({ nameError: "Container name is required" });
```
Missing translations `_("")`

strzinek replied:
Thanks, fixed in commit https://github.com/cockpit-project/cockpit-podman/commit/0c3706e1009fabc1fee2a17b247175993469a9c4

jelly commented on src/app.jsx:
```diff
@@ -367,6 +367,7 @@ class Application extends React.Component {
         case 'unmount':
         case 'unpause':
+        case 'rename': // rename event is available starting podman v4.1; until then the container does not get refreshed after renaming
             this.updateContainerAfterEvent(event.Actor.ID, system);
```
I've tested this on an older podman and it's a bit annoying that there is no visual feedback from the rename action. I think we should do something about that before merging the PR.

When /rename returns 204 did the rename already happen? If so we can manually refresh the container status I suppose until everything is podman 3.4

strzinek replied:
I agree and I would refresh it manually only if someone helped me how to call Application instance method correctly from handler code and not breaking architecture? I am new to react.

strzinek replied:
I figured out how to refresh the status of renamed container manually, it should be fixed now.

garrett commented:
- I can't rename running containers. I would have assumed they would be stopped and restarted, but still would be able to be renamed. (Show rename, show modal with a message at the top that mentions that the container will be renamed and restarted, then make the action "Rename and restart".)
- Hitting enter reloads the page. The default form action needs to be caught so it does the right thing, instead of doing an HTTP submit.
- Escape doesn't close the modal.
- The error is in the wrong place. It should, ideally, be directly underneath the field, not an alert. It should also be shorter and to the point.
   - https://www.patternfly.org/v4/components/form/design-guidelines#errors-and-validation
- When a string doesn't match the regular expression, it should say something about invalid characters... or it should just filter the characters like the Machines page does for naming. (Where invalid ones are stripped and spaces are changed to dashes or underscores, whichever is supported, unless spaces are simply supported in the names... then they shouldn't be stripped.) It should _not_ be an alert.

![image](https://user-images.githubusercontent.com/10246/164505793-96130a5a-a70e-4960-a877-67b2e6b67879.png)

garrett commented:
> About renaming running containers: I suppose it could be implemented as you assumed, but I tried to follow the use cases of podman itself.

Yeah. I just found it surprising. I think we'll probably want to implement it, but as a follow-up, not in this PR, as it's more complex.

> It would get even more complicated if container would have been paused - in such case it would have to be resumed first and stopped to rename the container. Or use force-stop in such cases? Let user decide in two steps then? I like to keep things simple.

We'd need to mention in the rename modal that that container will be stopped and restarted after the rename for running and paused containers.

These are a lot of things to think about, and I think it reinforces that this should be handled in a follow-up PR.

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,97 @@
+        };
+        setNameError(null);
+        setDialogError(null);
+        client.renameContainer(props.container.isSystem, props.container.Id, renameData)
```
Please do this inline, with `renameData` → `{ name }`; more succinct, and avoids the indirection with the single-use constant.

`{ x }` is JS shorthand for `{ x: x }`, BTW, as it's a very common use case.

martinpitt commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,97 @@
+        client.renameContainer(props.container.isSystem, props.container.Id, renameData)
+                .then(() => {
+                    props.onHide();
+                    // This is a workaround for missing API rename event in Podman.
+                    // It should be fixed in Podman v4.1, then next line and reference from props can be removed
+                    props.updateContainerAfterEvent(props.container.Id, props.container.isSystem);
```
Very good! :clap: A clean and documented workaround. Can you please add a "HACK:" to the comment? This is our standard way to mark such workarounds, easily greppable.

strzinek replied:
Added.

martinpitt commented on test/check-application:
```diff
@@ -1913,6 +1913,36 @@ class TestApplication(testlib.MachineCase):
+        self.performContainerAction(container_name, "Rename")
+
+        # the container name should be in the "Rename container" header
+        self.assertIn(container_name, b.text("#pf-modal-part-1"))
```
Please use `b.wait_in_text("#pf-modal-part-1", container_name)` for this, both for consistency and also to ensure that there are no race conditions with async rendering.

martinpitt commented on test/check-application:
```diff
@@ -1913,6 +1913,36 @@ class TestApplication(testlib.MachineCase):
+
+        container_sha = self.execute(auth, f"podman inspect --format '{{{{.Id}}}}' {container_name_new}").strip()
+        self.waitContainer(container_sha, auth, name=container_name_new, image='busybox:latest', cmd='sh',
+                           state='Configured', owner="system" if auth else "admin")
```
Well done on the test! Can you please now start the container, and pin down the behaviour for running containers? Judging by the code, the intention is that there should not be a "Rename" action, so the test should ensure that the kebab does not have that menu entry any more. Then, if at some point running container names are supported/get implemented, we have a reminder to update this.

strzinek replied:
Added.

strzinek replied:
>     * When a string doesn't match the regular expression, it should say something about invalid characters... or it should just filter the characters like the Machines page does for naming. (Where invalid ones are stripped and spaces are changed to dashes or underscores, whichever is supported, unless spaces are simply supported in the names... then they shouldn't be stripped.) It should _not_ be an alert.

I added a regexp validator for name, have a look. Machines names can now have non-ASCII characters, unlike Containers.

martinpitt reviewed (changes requested):
Nice progress! A few more small things, but well done!

strzinek commented:
@garrett wrote:
>     * When a string doesn't match the regular expression, it should say something about invalid characters... or it should just filter the characters like the Machines page does for naming. (Where invalid ones are stripped and spaces are changed to dashes or underscores, whichever is supported, unless spaces are simply supported in the names... then they shouldn't be stripped.) It should _not_ be an alert.

I added a regexp validator for name, have a look. Machines names can now have non-ASCII characters, unlike Containers.

garrett reviewed (changes requested):
It's looking nice.

Can we drop the `[ ]` surrounding the characters? It looks odd — ideally, it'd be some kind of `kbd` element which would make it styled similar to `pre` (but is inline instead). However, I know that having elements in translated strings can be tricky. I'd suggest just removing the squared off braces and show the characters as-is... and perhaps in parenthesis?

It's also missing commas separates after the text and each character.

"Invalid characters. Name must only contain letters, numbers, and certain punctuation (_ . -)."

![image](https://user-images.githubusercontent.com/10246/166248707-5850c97b-6913-498c-8fdb-e5f46b7f7835.png)

garrett commented:
I tried reviewing this today, but kept hitting a strange error where containers wouldn't load. I tried rebooting my system, switching between versions of Fedora (you can do that pretty easily with Silverblue), and so on.

Podman couldn't find a container and was just very, very upset. Even when I finally gave up and ran `podman system reset`, it couldn't remove some related file. I had to `rm -rf ~/.local/share/containers/` afterward. I nuked everything and started over. 

Do you think this could possibly be related to renaming and that I hit some sort of bug when testing this earlier?

I just tried renaming a container and it seemed to work. But running it fails with:

```
Failed to start container why_hello_there_
plugin type="bridge" failed (add): failed to find plugin "bridge" in path [/usr/local/libexec/cni /usr/libexec/cni /usr/local/lib/cni /usr/lib/cni /opt/cni/bin]
```

![Screenshot 2022-05-03 at 17-45-57 Podman containers - garrett@Bolt](https://user-images.githubusercontent.com/10246/166488198-0075efb0-3272-48ea-b881-02ab60a69373.png)


Is Podman's rename buggy? Or could it be code in the PR? Or both? This seems really suspicious — both the corruption earlier and now having a problem running the container.

Otherwise, the UI is pretty good. I think we still probably want to convert spaces to dashes or underscores instead of just throwing an error with a space. Otherwise, it's good UI-wise. However, if Podman is really causing problems when renaming containers, I think that might negate this entire PR, sadly, until Podman fixes its bugs (_if_ it is buggy with renaming).

garrett commented:
After renaming another container, I got an error about an improper container state:

![Screenshot 2022-05-03 at 17-51-12 Podman containers - garrett@Bolt](https://user-images.githubusercontent.com/10246/166488978-73f974cc-5a57-452c-9225-2e5c943f798c.png)

jelly commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,100 @@
+    const handleInputChange = (targetName, value) => {
+        if (targetName === "name") {
+            setName(value);
+            if (/^[a-zA-Z0-9][a-zA-Z0-9_\\.-]*$/.test(value)) {
```
If it's empty, it shows the wrong message:

![image](https://user-images.githubusercontent.com/67428/166490456-003bdd30-af7d-4908-aee8-474d9b71c549.png)

jelly commented on test/check-application:
```diff
@@ -1913,6 +1913,40 @@ class TestApplication(testlib.MachineCase):
+        b = self.browser
+        container_name = "rename"
+        container_name_new = "rename-new"
+        
```
This line has some extraneous whitespaces.

jelly commented on test/check-application:
```diff
@@ -1913,6 +1913,40 @@ class TestApplication(testlib.MachineCase):
+        container_sha = self.execute(auth, f"podman inspect --format '{{{{.Id}}}}' {container_name_new}").strip()
+        self.waitContainer(container_sha, auth, name=container_name_new, image='busybox:latest', cmd='sh',
+                           state='Configured', owner="system" if auth else "admin")
+        
```
This seems to have some extraneous whitespaces.

strzinek replied:
Fixed.

jelly reviewed (changes requested):
Triggered the tests, I think we still need to handle older podman versions without rename support. So let's see what the bots think about it.

jelly commented:
```
test/check-application:1926:1: W293 blank line contains whitespace
test/check-application:1946:1: W293 blank line contains whitespace
test/check-application:1951:1: E305 expected 2 blank lines after class or function definition, found 1
```
For the CI to run please fix these issues and I'll re-trigger the bots.

garrett commented:
> handle older podman versions without rename support

Don't show rename. I think we should only support rename where Podman does.

It looks like some form of rename is in Podman 3.0 (2021-02-11) and above, according to the history of the first commit @ https://github.com/containers/podman/commit/997de2f8e9e5453a99108bde012aa6c41d7323ec

[The release notes for 3.0](https://github.com/containers/podman/releases/tag/v3.0.0) say:

- Added the `podman rename` command, which allows containers to be renamed after they are created.

strzinek commented on src/ContainerRenameModal.jsx:
```diff
@@ -0,0 +1,100 @@
+                        isDisabled={nameError}
+                        onClick={handleRename}>
+                    {_("Rename")}
+                </Button>{' '}
```
I followed pattern used in ContainerDeleteModal.jsx. I assume I can just remove the spacing.

strzinek commented:
> Podman couldn't find a container and was just very, very upset. Even when I finally gave up and ran `podman system reset`, it couldn't remove some related file. I had to `rm -rf ~/.local/share/containers/` afterward. I nuked everything and started over.
> 
> Do you think this could possibly be related to renaming and that I hit some sort of bug when testing this earlier?
> 

I have never seen this. I had no problems running containers after renaming them. The code I wrote is calling standard Podman API, so if there is a problem, this is definitely bug on Podman side.

jelly commented:
> > > handle older podman versions without rename support
> > 
> > 
> > Don't show rename. I think we should only support rename where Podman does.
> 
> Rename action showed only for v3+ now.

Ok, running the tests, let's see what they make of it :)

Thanks for the updates!

strzinek commented:
@jelly :
> Please resolve this, then I can trigger the tests:
> 
> ```
> python3 -m pycodestyle --max-line-length=195 test/check-application # TODO: Fix long lines
> ```

Fixed, thanks.

martinpitt commented:
/packit build

martinpitt commented on src/Containers.jsx:
```diff
@@ -223,8 +230,16 @@ const ContainerActions = ({ container, onAddNotification, version, localImages }
                           onClick={() => startContainer()}>
                 {_("Start")}
-            </DropdownItem>
+            </DropdownItem>,
```
Why the trailing comma here? previous stanzas don't use it either, and it looks like an error. This isn't a list.

martinpitt commented on test/check-application:
```diff
@@ -1913,6 +1913,40 @@ class TestApplication(testlib.MachineCase):
+        self.execute(auth, f"podman inspect --format '{{{{.Id}}}}' {container_name_new}").strip()
+        self.waitContainerRow(container_name_new)
+
+        # test if rename is hidden for running container. This should change with rename events fixed in Podman.
```
Getting an event for a rename seems independent from supporting renames for running containers. Is that comment actually true? I thought events were already fixed in podman 4.1.

jelly replied:
I did some digging, and indeed the rename event is added in 4.1 https://github.com/containers/podman/releases/tag/v4.1.0

About renaming a running container, this works surprisingly well?

```
[root@fedora-36-127-0-0-2-2201 ~]# podman ps
CONTAINER ID  IMAGE                          COMMAND     CREATED       STATUS            PORTS       NAMES
359ac580010e  quay.io/libpod/busybox:latest  /bin/sh     1 second ago  Up 2 seconds ago              foo
[root@fedora-36-127-0-0-2-2201 ~]# podman rename foo bar
[root@fedora-36-127-0-0-2-2201 ~]# podman ps
CONTAINER ID  IMAGE                          COMMAND     CREATED        STATUS             PORTS       NAMES
359ac580010e  quay.io/libpod/busybox:latest  /bin/sh     9 seconds ago  Up 10 seconds ago              bar
[root@fedora-36-127-0-0-2-2201 ~]# podman -v
podman version 4.0.2
```

jelly replied:
And also on 3.0.1?

```
root@debian:~# podman ps
CONTAINER ID  IMAGE    COMMAND  CREATED       STATUS            PORTS   NAMES
cdc59dfb3e96  busybox  /bin/sh  1 second ago  Up 2 seconds ago          foo
root@debian:~# podman rename foo bar
root@debian:~# podman ps
CONTAINER ID  IMAGE    COMMAND  CREATED        STATUS            PORTS   NAMES
cdc59dfb3e96  busybox  /bin/sh  7 seconds ago  Up 7 seconds ago          bar
root@debian:~# podman -v
podman version 3.0.1
```

martinpitt reviewed (changes requested):
Some tiny issues, but LGTM overall. I started the tests. Many thanks!

garrett reviewed (changes requested):
I think this is almost there. Thanks for answering about the bugs I found (which really could be — and probably are — specific to my system).

1. Inline alerts should be at the top of modals, according to PatternFly guidelines.

   ![image](https://user-images.githubusercontent.com/10246/167632365-bfdc7909-5c1d-4eee-b995-83eb8bc0b423.png)

2. (_Optional; could be follow-up_) If we could focus the entry when the modal comes up, that would be nice. 

3. (_Optional; could be follow-up_) If we can rename running containers, then that'd be great to have in here.

Reference for errors in form validations on PatternFly's website:
https://www.patternfly.org/v4/components/form/design-guidelines/#form-sections:

> When a form field submission results in an error, let users know as soon as possible and as close as possible to the error. Default to presenting error states on a form using field level errors whenever possible.
>
> In some use cases, you may choose to use inline errors at the top of a form to emphasize errors within it. These inline errors should always be a supplemental component used in addition to field level errors.

...and then, a little further down on the page:

> For validation on submission, supplement field level errors with an [inline alert](https://www.patternfly.org/v4/components/alert/design-guidelines#inline-alerts) to make it simple for the user to quickly identify and address areas that need attention. The inline alert should be placed below the form’s header and description, and it should equal the total width of the form.

Since the alert's details are rather long, we have the option of either using a more specific error message (about the container already existing) and using an expandable alert (to hide the extra information, if we do think it's necessary, even with a more precise error message) so that the alert isn't so huge and doesn't push everything down by default. I think it just needs to have an `isExpandable` attribute in the `<Alert>` component.

More info on expandable alerts: https://www.patternfly.org/v4/components/alert#expandable

strzinek commented:
@garrett wrote:
>     1. Inline alerts should be at the top of modals, according to PatternFly guidelines.
Changed.

>     2. (_Optional; could be follow-up_) If we could focus the entry when the modal comes up, that would be nice.
When I tried to auto focus on input element in modal rename form, this error was reported during make: `error  The autoFocus prop should not be used, as it can reduce usability and accessibility for users  jsx-a11y/no-autofocus`

>     3. (_Optional; could be follow-up_) If we can rename running containers, then that'd be great to have in here.
Done. I added conditions for versions 3.0.1+ for renaming running/paused containers and reverted test.

jelly reviewed (approved):
Everything looks good to me, except the one extraneous comma

strzinek commented:
> Everything looks good to me, except the one extraneous comma

I deleted all of them, even the ones that were there originally.

jelly commented:
> > Everything looks good to me, except the one extraneous comma
> 
> I deleted all of them, even the ones that were there originally.

Oh that was wrongly suggested by me, the commas should be there for pushing the elements on the list. Sorry about that.

strzinek commented:
> Oh that was wrongly suggested by me, the commas should be there for pushing the elements on the list. Sorry about that, if you re-push please squash everything into one commit.

Should I squash the whole PR?

jelly commented:
> > Oh that was wrongly suggested by me, the commas should be there for pushing the elements on the list. Sorry about that, if you re-push please squash everything into one commit.
> 
> Should I squash the whole PR?

Yes, please the whole feature should be one commit

strzinek commented:
> Yes, please the whole feature should be one commit

But isn't this done while merging the PR to main?

martinpitt commented:
Build fails:
```
ERROR in /tmp/tmp0mvkw23n/src/Containers.jsx
  196:12  error  Parsing error: Adjacent JSX elements must be wrapped in an enclosing tag. Did you want a JSX fragment <>...</>? (196:12)
```

As long as the result is just a single commit, we can squash&merge, yes. We don't usually do this, but if you prefer to work with such a huge "fixup" commit list, you can keep doing that.

garrett reviewed (approved):
Looks good to me from a design standpoint! :+1: Thanks for adding rename support!

jelly commented:
It seems you have set up some sort of pull request syncing mechanism when something is merged in main, this doesn't work well with our CI as it rebases automatically which it can't.

I've pushed your old commit and rebased it locally, so now the tests run and it's good to be merged when everything is green.

jelly commented:
Thanks for creating the PR and the follow up fixes :tada:

strzinek commented:
I haven't had time to pay much attention to it today. Thank you for sorting it out.

jelly commented:
> I haven't had time to pay much attention to it today. Thank you for sorting it out.

No worries, I am happy it's merged. It's a nice feature to have!
